<!-- Fix images in UNIX section so that they use 1005 instead, to match the new text. -->

<!-- Explain importance of scripting versus click-and-drag and copy-pasting. Add section about useful keyboard shortcuts. -->

<!-- Add material about file organization. -->

<!-- Find a place to mention all the high quality sources listed below. Maybe with a webshot of their title pages?  -->

<!-- Happy Git for the R User. -->
<!-- Anything associated with RStudio. -->
<!-- [The UNIX Workbench](https://seankross.com/the-unix-workbench/), but with all the work done form the terminal window in RStudio.  -->
<!-- [What They Forgot to Teach You About R](https://rstats.wtf/) -->
<!-- https://usethis.r-lib.org/articles/articles/usethis-setup. Might just pull out the key sections of this. We can't just link to it since some of it --- like devtools --- is too confusing. -->
<!-- https://merely-useful.github.io/py-rse/ -->

<!-- Consider making the **fs** package a major part of the presentation. This helps avoid dealing with cross-OS issues ourselves. Also teaches students that there is more than one way to interact with the file system. -->

<!-- Add some optional videos, especially the ones from RStudio which give nice tours of RStudio stuff. -->

<!-- Buttons we press should be bolded, not wrapped with backticks. -->


# Tools {-}

The chapter will be broken into the following sections. Read whichever ones are relevant.

* [Absolute and relative file paths](#absolute-and-relative-file-paths)
* [Working with the terminal](#working-with-the-terminal)
* [Git, GitHub, and RStudio](#git-github-and-rstudio)
* [PDF](#pdf)
* [Style guide](#style-guide)  
* [How to use Rpubs](#how-to-use-rpubs)  
* [How to get help](#how-to-get-help)
* [How to make a table](#how-to-make-a-table)

## Absolute and relative file paths

<!-- DK: Move this section to tools.Rmd. -->
<!-- AG: I added this as another section because it's important enough to stand on its own. This also serves as a great way to introduce the Terminal.-->

When you read data into R, you first need to tell R where that data lives. Most times, the data will be in a file. The file could live on your computer (local) or somewhere on the internet (remote). 

The place where the file lives on your computer is called the "path". You can think of the path as directions to the file. The path includes both the location of the file and the name of the file itself. There are two kinds of paths: relative and absolute. A *relative* path describes the location of the file relative to where you currently are on the computer. An *absolute* path is where the file is in respect to the base (or root) folder of the computer's filesystem. Absolute paths always start with a forward slash, a "/". 

Consider a file called `report.csv`. Read the file using a relative path:

```
x <- read_csv("data/report.csv")
```

Read `report.csv` using an absolute path:

```
x <- read_csv("/home/preceptor/desktop/projects/data/report.csv")
```

To ensure your code can be run on a different computer, you should use relative paths. An added bonus is that it's also less typing! This is because the absolute path of a file (the names of folders between the computer's root `/` and the file) isn't usually the same across different computers. For example, suppose Fatima and Jayden are working on a project together on the `report.csv` data. Fatima's file is stored at 

`/home/Fatima/files/report.csv`, 

while Jayden's is stored at 

`/home/Jayden/files/report.csv`.
 
Even though Fatima and Jayden stored their files in the same place on their computers, the absolute paths are different due to their different usernames. If Jayden has code that loads the `report.csv` data using an absolute path, the code won't work on Fatima's computer. But the relative path from inside the `files` folder (`files/report.csv`) is the same on both computers; any code that uses relative paths will work on both!

<!-- Include a section about how you're supposed to put \ whenever there's a space, since 90% of people will have a space in their file name. However, because we don't have a section about it, we get a lot of confusion about why somebody's thing isn't working. Fixed - AG -->

One important part of using paths is recognizing when there are spaces in the file name. For example, trying to use the path

`/home/preceptor/desktop/projects/important data/report.csv`

does not work because it includes a space the file name, with "important data". Instead, the path

`/home/preceptor/desktop/projects/important\ data/report.csv`

is valid and allows you to access the `report.csv` data. This is because the `\` tells the computer to treat the next character only as a character instead of something special. For example, the space character is normally used to show a break in the code, so the computer treats it as a break point whenever it sees a space. However, `\ ` tells the computer that the space isn't a break point and is instead part of the file path, solving the issue.

<!-- DK: Video includes a bunch of good stuff, like ~, ., .., and /. All that should be mentioned here and connected to the terminal tutorial. -->

See this video for another explanation: 

<iframe width="840" height="473" src="https://www.youtube.com/embed/ephId3mYu9o" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

*Source: [Udacity course "Linux Command Line Basics"](https://www.udacity.com/course/linux-command-line-basics--ud595)*


## Working with the terminal {-}

<!-- DK: Could rework everything to use this:  -->
<!-- https://merely-useful.github.io/py-rse/bash-basics.html -->

The Terminal is a very powerful window because it allows you to interact with your computer's filesystem directly, and it uses file paths to find the files that you want to interact with.

Let's open up the Terminal tab on the left window and start learning how to use the Terminal.

### `pwd`: Working directory {-}

The first question you may have about working in the Terminal might be: If I can't see the folders, how do I know where I am? Well that's a great place to start learning the Terminal. To see what our current folder is, we type `pwd` (**p**rint **w**orking **d**irectory):

```{r, echo = FALSE}
knitr::include_graphics("other/images/pwd.png")
```

We are currently in a directory (or folder) called `Yao`, which is itself in a directory named `/Users`. The forward slash in front of "Users" tells is that this directory is at the lowest possible level.

### `ls`: Seeing items in the directory {-}

To see the items in our current folder, we use the command `ls` (**l**i**s**t). Type `ls` in the terminal and hit return/enter. You should see something like this:

```{r, echo = FALSE}
knitr::include_graphics("other/images/ls.png")
```

Notice that this lists exactly the same items as the bottom right window in RStudio. The Terminal is just another way to interact with your computer's filesystem. Anything you can do normally with your mouse/trackpad, like opening a folder, you can also do in the Terminal.

```{r, echo = FALSE}
knitr::include_graphics("other/images/ls_icons.png")
```

### `cd`: Changing directories {-}

To move from one directory to another, we use `cd` (**c**hange **d**irectory). We'll be using `cd` to change into the Desktop folder.

To change into the Desktop directory, we type `cd Desktop/`. A helpful hint, after you type the first few letters of a folder or file name, you can hit tab and the computer will auto complete the name. Try it! Type `cd Desk` and then hit tab to auto complete the name!

```{r, echo = FALSE}
knitr::include_graphics("other/images/cd_desktop.png")
```

If you type `ls` again, you can see all the item on your Desktop listed. 

To go back to the previous folder (aka the directory above), we can type `cd ..` The two periods represent one level above. You can see this hierarchy in this view on a Mac:

```{r, echo = FALSE}
knitr::include_graphics("other/images/directory_hierarchy.png")
```

### `mkdir` and `rmdir`: Make and remove a directory {-}

Now that we're in the Desktop folder, let's get set-up to stay organized for Gov 1005. Staying organized is critical when working with many data projects. So, using `mkdir Gov-1005` (make directory) we can create a folder exclusively for Gov 1005 like so:

```{r, echo = FALSE}
knitr::include_graphics("other/images/mkdir.png")
```

Now, when we type `ls`, we can see our new folder created. Note that we used a hyphen between Gov and 1005. This is because the Terminal can't recognize spaces unless you put `\` before it, like so: `mkdir Gov\ 1005`. *Never use spaces or other weird characters in file or directory names.*

To remove a folder, use `rmdir` (remove directory). We won't be using this right now because we don't need to remove anything.

### `touch:` Creating files {-}

In order to experiment with the next few commands in the Terminal, we'll need a test file. Type `touch text.txt` to create a test file. 

```{r, echo = FALSE}
knitr::include_graphics("other/images/touch.png")
```

And, of course, we can see that the `test.txt` file has been created using `ls`.

### `mv`: Moving files {-}

Oh no! We created our `test.txt` file, but it should be in our `Gov-1005` folder, right now it's on the desktop. This happened because while we created the `Gov-1005` folder using `mkdir`, we forgot to move into it by using `cd Gov-1005/`. But no worries, we can move the file to that folder using `mv`:

```{r, echo = FALSE}
knitr::include_graphics("other/images/mv.png")
```

When using `mv` the first thing you type after `mv` is the file you want to move. The next thing is the location where you want to move it to. In our case we want to move `test.txt` to `Gov-1005/`, so we type `mv test.txt Gov-1005/`. After we do this, we can use `cd` to enter the `Gov-1005` folder and then use `ls` to see that our `test.txt` file successfully was moved into the `Gov-1005` directory.

```{r, echo = FALSE}
knitr::include_graphics("other/images/mv_confirm.png")
```

### `cp`: Copying files {-}

Copying files is very similar to moving files in the Terminal. Using the previous example, if we wanted to copy test.txt into the Gov-1005 folder but not delete the original test.txt file, we just replace `mv` with `cp` (copy paste):

```{r eval=FALSE}
cp test.txt Gov-1005/
```


### `rm`: Removing files {-}

Ok, we are at the last Terminal command that this book will be teaching you. So, we're done with this test.txt file. Let's remove it with `rm` (remove):

```{r, echo = FALSE}
knitr::include_graphics("other/images/mv_confirm.png")
```

Make sure you are the in Gov-1005 folder before you type `rm test.txt`! Using `ls`, we can see that our test file is now gone.

Congrats! You are now able to do most basic tasks with the Terminal! If you want to learn more Terminal commands, check out Sean Kross's [The Unix Workbench](https://seankross.com/the-unix-workbench/).


## Git, GitHub, and RStudio {-}

This next section focuses on connecting GitHub with RStudio using Git. Why do we care about GitHub? Think of it as a Google Drive for all your R code and projects. If your computer blows up, GitHub will save all your R work just as Google Drive saves your paper.

### Installing Git  {-}

The first step to using GitHub is installing Git on your computer. But first, you may already have Git installed on your computer. To check, go to your Terminal and type `git --version`. If you already have Git, this command will return the Git version you have installed. If you get an error, you can download and install git [here](https://git-scm.com/downloads). 

### GitHub accounts {-}

<!-- DK: Add Happy advice on  this and other topics? -->

After installing Git, you'll need a [GitHub](https://github.com/join) account. This is like a Google account. However, the one difference with GitHub is that your account is visible to the public. You want to pick a name carefully. It should be professional since you will be sending potential employers a link to your GitHub account in the near future. Check out some former Gov 1005 students' GitHub profiles for inspiration: 

* [Evelyn Cai](https://github.com/caievelyn)

* [Jessica Edwards](https://github.com/jessie9111)

* [Beau Meche](https://github.com/BeauMeche)

Once you have a GitHub account, you are ready to connect Git and RStudio to this account. Type the following two commands in your Terminal pane. Replace `Your Name` with your name and `your@email.com` with the email you used to sign up for GitHub.

```{eval=FALSE, echo=TRUE}
git config --global user.name "Your Name"
git config --global user.email "your@email.com"
```

### GitHub repositories {-}

You are now ready to create a GitHub repository (repo). The GitHub repo is similar to a Google Drive folder. To make your first repo, make sure you are signed in and then go to the GitHub [homepage](https://github.com/) and click on the green `new` button on the left.

```{r, echo = FALSE, out.width="80%", fig.align='left'}
knitr::include_graphics("other/images/new_repo.png")
```

You will then want to choose a good name for the repo and add a brief description. Here we will use `productivity`. You can choose to make the repo public or private, but we recommend that you only make a repo public if it is important for the world to see. This keeps your public GitHub profile clean and professional. This repo should probably be private. Let's also add a README file for our repo. This is a document where you can add more information.

```{r, echo = FALSE}
knitr::include_graphics("other/images/repo_details.png")
```

You now have your first repo on GitHub. The next step will be to download it to your computer --- a process often known as "cloning" --- and start editing and syncing using Git. To do this, we'll need to copy the link to the repo and then use it in RStudio. Once again, the green button is your friend. Click on it and then copy the link shown. You can use the clipboard button on the right to automatically copy it.

```{r, echo = FALSE}
knitr::include_graphics("other/images/repo_clone.png")
```

### Connecting GitHub to RStudio {-}

We are now ready to connect your `productivity` repo to RStudio. With the link to the productivity repo copied, we can go back to RStudio and begin with a new project. Go to File, then New Project:

```{r, echo = FALSE}
knitr::include_graphics("other/images/new_project.png")
```

Next, you'll need to go through these steps to create the project: Version Control to Git to paste your link from GitHub and click Create Project.

```{r, echo = FALSE}
knitr::include_graphics("other/images/new_project.gif")
```

Congrats! You've linked your `productivity` repo to RStudio. Note that Github will ask you for a location in which to place this and other projects. *We recommend creating a folder on your desktop called "projects" and placing all your RStudio projects there.* Don't just scatter them across your computer in a mess. There will be dozens of them. Be organized!

<!-- DK: Create a gif which walks through the above process. -->


### Updating `.gitignore` {-}

The first thing you should always do when working with a new repo is updating the `.gitignore` file. You can open this file from the bottom right window under the Files tab. This file includes all the files that you don't want to be uploaded to GitHub. This can come in handy when you are working with big datasets or files with private information. In our case, we want to add the `productivityl.Rproj` file to the `.gitignore` list. 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rproj_git.png")
```

This file is your private project file and usually you don't want this uploaded to GitHub. So, in `.gitignore`, you'll want to add `*.Rproj` The * tells your computer that we want to prevent all files ending in `.Rproj` from being uploaded. We could also just add `productivity.Rproj`.

```{r, echo = FALSE}
knitr::include_graphics("other/images/rproj.png")
```

Save the `.gitignore` file and you should see the `productivity.Rproj` file disappear from your Git tab in the top right window. If you don't see any changes, click on the refresh button in the upper left. 

The symbols in the Git tab are part a "conversation" between you and Git. The "?" is Git's way of saying: "There is a new file here. What do you want to do with it?" Adding a line to the `.gitignore` is your way of replying "Ignore that file."

```{r, echo = FALSE}
knitr::include_graphics("other/images/updated_gitignore.png")
```

### Commit and Push {-}

Now that we've updated our `.gitignore` file, we want to upload this new version to GitHub. To do this, first select the `.gitignore` file and then click on the `Commit` button in the Git window:

```{r, echo = FALSE}
knitr::include_graphics("other/images/commit.png")
```

This will open a new window where you will write a commit message. This message is a very short note on what you're adding/changing in the repo. In our case, we've updated the `.gitignore` so let's write just that:

```{r, echo = FALSE}
knitr::include_graphics("other/images/commit_push.png")
```


Press `commit`. This is your way of telling Git "Yes these are the files I want to upload. I'm committed." Next, press **Push**. This pushes or uploads the files to GitHub. (You can probably guess what pull does, but we won't be using that yet)

Now, if you go to your GitHub repo and refresh the page, you can see that the `.gitignore` file has been uploaded with your commit message:

```{r, echo = FALSE}
knitr::include_graphics("other/images/github_updating.gif")
```

Congrats! You just uploaded your first file to GitHub.

One tricky aspect is the caching of your Github ID and password. Most likely, you had to type these things in when you did your first push. Doing so was not too bad. And, after all, Github needs to know who you are, otherwise other people could mess with your repo. But you will be doing hundreds of commits/pushes. You don't want to type in your ID/password each time! Follow [these instructions](https://usethis.r-lib.org/articles/articles/git-credentials.html). Key steps:

1) Turn on two-factor authentication for your GitHub account under [Settings -> Security](https://github.com/settings/security).

2) Create a token:

```{r, eval = FALSE}
usethis::create_github_token()
```

This will, after logging in, bring you back to Github. Accept the defaults and press the **Generate token** button at the bottom. (You may need to change the **Note** if you have generated tokens before.) Copy the token which has been created. It will look something like:

8be3e800891425f8462c4491d9a4dbb5b1c1f35c

Then, issue this R command:

```{r, eval = FALSE}
gitcreds::gitcreds_set()
```

Provide your token. After you start a new RStudio instance, Github should not ask you for your login/password again. Or it might just ask one more time. Seek help if this does not work.

<!-- DK: Could use a gif for the above as well. -->


[*Happy Git and GitHub for the useR*](https://happygitwithr.com/) is the best source for when Git or Github problems arise.

## PDF {-}

Generating PDF files from RStudio is both easy and hard. It is easy because R markdown is designed to produce files in a variety of output formats, [including PDF](https://bookdown.org/yihui/rmarkdown/pdf-document.html). It is hard because, for RStudio to make PDF files, your computer set up must be set up with a [LaTeX](https://en.wikipedia.org/wiki/LaTeX) installation. You have four options:

1) Making PDF files may just "work," especially if you are using a Mac. Give it a try!

2) If it doesn't just work, we *strongly* recommend using the **[tinytex](https://yihui.org/tinytex/)** R package. First, install the R package.

```{r, eval=FALSE}
install.packages('tinytex')
```

Second, use the R package to install the underlying LaTeX distribution.

```{r, eval=FALSE}
tinytex::install_tinytex()
```

Depending on your operating system, this may not work. But there should be an error message providing further instructions. Follow those instructions.

Restart R and everything should just work. 

3) You can just generate an html file, open it in Chrome, select *Print . . .* from the drop-down menu. You will get a pop-up window. Click the down arrow to the right of *Destination* and choose *Save as PDF* in the drop-down menu. You’ll see a preview. Choose the *Save as PDF* option. This is not a convenient workflow but, if disaster strikes and the problem set is due in 10 minutes, it is a reasonable option.

4) You can install a full LaTeX installation yourself. Good luck! Don't come to us for help.

## Style guide {-}

Much of this material comes from the *[Tidyverse Style Guide](https://style.tidyverse.org/index.html)*.  We will take off points on work submitted which violates these guidelines. *In extremis*, you may go against this advice, if you add a code comment in your work explaining your decision to do so.

### Comments {-}

Include comments in your code. Easy-to-understand chunks of code should not have comments. The code is the comment. But other code will merit many, many lines of comments, more lines than the code itself. In a given file, *you should have about as many total lines of comments as you have lines of code*.

Make your comments meaningful. They should not be a simple description of what your code does. The best comments are descriptions about why you did what you did and which other approaches you tried or considered. (The code already tells us what you did.) Good comments often have a “Dear Diary” quality: “I did this. Then I tried that. I finally chose this other thing because of reasons X, Y and Z. If I work on this again, I should look into this other approach.” Because of this, the structure is often a paragraph of comments followed by several lines of code. 

Each line of a comment should begin with the comment symbol (a "hash") followed by a single space: `# `. Code comments must be separated from code by one empty line on both sides. Format your code comments neatly. Ctrl-Shift-/ is the easiest way to do that. Name your R code chunks, without using weird characters or spaces. `download_data` is a good R code chunk name. `Plot #1` is not.   





Spelling matters. Comments should be constructed as sentences, with appropriate capitalization and punctuation. 

### Graphics {-}

Use captions, titles, axis labels and so on to make it clear what your tables and graphics mean.

Anytime you make a graphic without a title (explaining what the graphic is), a subtitle (highlighting a key conclusion to draw), a caption (with some information about the source of the data) and axis labels (with information about your variables), you should justify that decision in a code comment. We (try to) always include these items but there are situations in which that makes less sense. Ultimately, these decisions are yours, but we need to understand your reasoning.

Use your best judgment. For example, sometimes axis labels are unnecessary. Read *[Data Visualization: A practical introduction](https://socviz.co/)* by Kieran Healy for guidance on making high quality graphics.

### Formating

#### Long Lines {-}

Limit your code to 80 characters per line. This fits comfortably on a printed page with a reasonably sized font.  When calling functions, you can omit the argument names for very common arguments (i.e. for arguments that are used in almost every invocation of the function). Short unnamed arguments  can also go on the same line as the function name, even if the whole function call spans multiple lines.


#### Whitespace {-}

`%>%` should always have a space before it, and should usually be followed by a new line. After the first step in the pipe, each line should be indented by two spaces. This structure makes it easier to add new steps (or rearrange existing steps) and harder to overlook a step.

```{r, eval=FALSE}
# Good

iris %>%
  group_by(Species) %>%
  summarize_if(is.numeric, mean) %>%
  ungroup() %>%
  gather(measure, value, -Species) %>%
  arrange(value)

# Bad

iris %>% group_by(Species) %>% summarize_all(mean) %>%
ungroup %>% gather(measure, value, -Species) %>%
arrange(value)
```

**ggplot2** code is handled in a similar fashion. All commands after the initial invocation of `ggplot()` are indented.

```{r, eval=FALSE}
# Good

diamonds %>% 
  ggplot(aes(x = depth)) +
    geom_histogram(bins = 100) +
    labs(title = "Distribution of Depth",
         x = "Depth",
         y = "Count")

# Bad

diamonds %>% 
ggplot(aes(x = depth)) +
geom_histogram(bins = 100) + labs(title = "Distribution of Depth",
         x = "Depth",
         y = "Count")
```



#### Commas {-}

Always put a space after a comma, never before, just like in regular English.

```{r, eval = FALSE}
# Good

x[, 1]

# Bad

x[,1]
x[ ,1]
x[ , 1]
```

#### Parentheses {-}

Do not put spaces inside or outside parentheses for regular function calls.

```{r, eval = FALSE}
# Good

mean(x, na.rm = TRUE)

# Bad

mean (x, na.rm = TRUE)
mean( x, na.rm = TRUE )
```


#### Infix operators {-}

Most infix operators (`=`, `==`, `+`, `-`, `<-`, `~`, et cetera) should be surrounded by one space.

```{r, eval = FALSE}
# Good

height <- (feet * 12) + inches
mean(x, na.rm = TRUE)
y ~ a + b


# Bad

height<-feet*12+inches
mean(x, na.rm=TRUE)
y~a + b
```

Other operators ---  like `::`, `:::`, `$`, `@`, `[`, `[[`, `^`, and `:` --- should never be surrounded by spaces.

```{r, eval = FALSE}
# Good

sqrt(x^2 + y^2)
df$z
x <- 1:10

# Bad

sqrt(x ^ 2 + y ^ 2)
df $ z
x <- 1 : 10
```


You may add extra spaces if it improves alignment of `=` or `<-`.

```{r, eval = FALSE}
list(total = a + b + c,
     mean = (a + b + c) / n)
```

Do not add extra spaces to places where space is not usually allowed.

### Messages/Warnings/Errors  {-}

R messages/warnings/errors should never appear in a submitted document. The right way to deal with these issues is to find out their cause and then fix the underlying problem. Students sometimes use “hacks” to make these messages/warnings/errors disappear. The most common hacks involve using code chunk options like `message = FALSE`, `warning = FALSE`, `results = "hide"`, `include = FALSE` and others. Don’t do this, in general. A message/warning/error is worth understanding and then fixing. Don’t close your eyes (metaphorically) and pretend that the problem doesn’t exist. There are some situations, however, in which, no matter what you try, you can’t fix the problem. In those few cases, you can use one of these hacks, but you must make a code comment directly below it, explaining the situation. The only exception is the “setup” chunk (included by default in every new Rmd) which comes with `include = FALSE`. In that chunk, no explanation is necessary, by convention.



## How to use Rpubs {-}

[Rpubs](https://rpubs.com/) provides a free hosting service for your R work. To use it:

1. Begin by creating a new repository on GitHub. Then clone it to your computer. We are calling the repository "rpubs_example." As before, put *Rproj in your `.gitignore` file. This is to prevent your private project file from being uploaded to GitHub.

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-1.png")
```

2. Start a new R Markdown file. Go to File --> New File --> R Markdown. For simplicity, leave the name "Untitled" and hit "OK." 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-2.png")
```

3. Save this file, again, as "Untitled" in your project directory. 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-3.png")
```

4. Knit. You should see the following. 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-4.png")
```

5. Notice a blue icon in the upper right-hand corner that reads "Publish." Click it. 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-5.png")
```


6. You will be asked whether you want to publish to RPubs or RStudio Connect. Choose RPubs. You will get a reminder that all documents you publish on RPubs are publicly visible. Click "Publish." 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-6.png")
```

7. This will take you to the RPubs website. You will need to create an account. Follow the steps as prompted. 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-7.png")
```

8. Add document details. Name your document. Add a meaningful slug -- otherwise you will end up with an ugly, long address you didn't choose and can't remember. You can leave the Description blank for simplicity of the exercise. 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-8.png")
```

9. Hit "Continue", et voilá! You have published your first document to Rpubs! 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-9.png")
```

10. There is one more important step. "rsconnect" contains  files specific to your computer that you do not want to push to GitHub. Therefore, as with .Rproj files before, we want to add the rsconnect folder to the `.gitignore` file. Click on `.gitignore`, add it there and hit "Save." You will see it disappear from your GitHub in the top right window. If you don't see any changes, hit the Refresh button in the top right corner. Since you've updated your `.gitignore` file, now is a good time to commit and push your changes to your GitHub repository. 

```{r, echo = FALSE}
knitr::include_graphics("other/images/rpubs-10.png")
```


## How to get help {-}

*The best data science superpower is knowing how to ask a question.* -- Mara Averick


### Searching for Help with R {-}

Google is your best friend. If you have a question about doing something in R, someone has probably had the same question and someone else has probably answered it online. [Stack Overflow](stackoverflow.com/) and [RStudio Community](https://community.rstudio.com/) are two of the best forums for finding and asking questions/solutions. Adding "R" and/or "tidyverse" in your keyword search helps find more relevant results to your exact question. Be as specific as possible in wording your question!

### Reproducible Examples {-}

If you don't find an answer to your question and you're still stuck, then you should ask your own question on these forums! In order to get the best response, sharing a reproducible example with the community allows others to easily start where you left off. The **reprex** package allows you to do this easily.

First, install the **reprex** package. Then, load the package.

```{r, eval=FALSE}
install.packages("reprex")
library(reprex)
```

<!-- DK: Switch this to use a built in data set. -->

Let's now look at some problematic code. The data set "murders" from the package "dslabs" provides murder statistics, as well as population counts, for all states. Suppose we want to calculate the rate of murders in each state per 100k residents (number of murders/population * 10^6). However, in the code below we have misspelled "population", resulting in this error: 

```{r, echo = FALSE, out.width="110%"}
knitr::include_graphics("other/images/reprex-2.gif")
```

To use **reprex**, highlight the code *and* the necessary packages. Then copy the highlighted code by pressing Ctrl/Cmd + c. The code is now in the clipboard. In the Console, type reprex() and hit Enter/Return. **reprex** then automatically creates a reproducible example and places it in your clipboard. 

```{r, echo = FALSE, out.width="110%"}
knitr::include_graphics("other/images/reprex-3.gif")
```

Now all you have to do now is go to your favorite R forum and paste your reproducible example into your question! Here is an example of how you would post it into Slack. 

```{r, echo = FALSE, out.width="110%"}
knitr::include_graphics("other/images/reprex-4.gif")
```

And here is how you would post it to RStudio community. 

```{r, echo = FALSE, out.width="110%"}
knitr::include_graphics("other/images/reprex-5.gif")
```

The first time you post to RStudio Community, be sure to keep the window open for a few seconds after you have pasted your question into the window. (Doing to will prevent your post from being sent to Moderation out of concern that you are a bot.)


## How to make a table {-}

**gt** is an R package for creating elegant tables. First, we'll create a **gt** summary table of some observations from the data. Second, we'll run a regression and display the outcome using **gtsummary**, a companion package to **gt** which specializes in presenting the results of statistical models.

If you want to learn more about **gt** check [this](https://themockup.blog/posts/2020-05-16-gt-a-grammer-of-tables/) fantastic guide. Go [here](https://gt.rstudio.com) for the official **gt** package website. See [here](https://www.danieldsjoberg.com/gtsummary/articles/tbl_summary.html) for an extensive guide to **gtsummary**.

Load the necessary libraries.

```{r setup, message=FALSE}
library(tidyverse)
library(primer.data)
library(gt)
```

We set `message=FALSE` in the above code chunk to avoid showing all the ugly notes when these libraries are loaded.

Let's pull some data which we will use in our table: 

```{r}
x <- trains %>%
  select(gender, income, att_end) %>%
  slice(1:5)
x
```

Create the simplest table with `gt()`, the key command:

```{r}
x %>% 
  gt()
```

&nbsp;

Now let's make this more professional. **gt** offers a variety of functions to add features like these^[source: https://gt.rstudio.com]: 

```{r, echo = FALSE}
knitr::include_graphics("other/images/gt_features.png")
```

You can add a title and subtitle using `tab_header()`: 

```{r}
x %>% 
  gt() %>%
   tab_header(title = "Enos Data Observations", 
              subtitle = "Gender, Income, and End Attitude from the Trains Data")
```

&nbsp;

By default, titles and other text can not be formatted. If you want formatting, you must wrap the character string in a call to `md()`, where *md* stands for (M)ark(d)own. For example, here is a bolded title.

```{r}
x %>% 
  gt()%>%
   tab_header(title = md("**Enos Data Observations**"), 
              subtitle = "Gender, Income, and End Attitude from the Trains Data")
```

&nbsp;

We can use `tab_spanner()` to add spanner columns. The `c()` argument takes in the variables that the spanner column will cover.

```{r}
x %>% 
  gt()%>%
   tab_header(title = md("**Enos Data Observations**"), 
              subtitle = "Gender, Income, and End Attitude from the Trains Data") %>% 
   tab_spanner(label = "Name of Spanner Column Here", c(gender, income))
```

From here on, our current table will not include a spanner column. If you wish to see more examples of spanner columns, go to [Chapter 4](https://ppbds.github.io/primer/rubin-causal-model.html#a-single-value-for-tau).

&nbsp;


You can change the column names using `cols_label()`:

```{r}
x %>% 
  gt()%>%
    tab_header(title = md("**Enos Data Observations**"), 
               subtitle = "Gender, Income, and End Attitude from the Trains Data") %>%
    cols_label(gender = "Gender",
               income = "Income", 
               att_end = "End Attitude")
```

&nbsp;

Use `tab_source_note()` to cite the source of the data or to create a caption. This function is not exclusively for providing a source --- though it's a handy way to do so --- and can be used to display any text you'd like:

```{r}
x %>% 
  gt()%>%
   tab_header(title = md("**Enos Data Observations**"), 
              subtitle = "Gender, Income, and End Attitude from the Trains Data") %>%
  cols_label(gender = "Gender",
             income = "Income", 
             att_end = "End Attitude") %>% 
  tab_source_note("Source: Ryan Enos")
```

&nbsp;

Using `md()` again, we can italicize the name of the Enos study in the caption:

```{r}
x %>% 
  gt()%>%
   tab_header(title = md("**Enos Data Observations**"), 
              subtitle = "Gender, Income, and End Attitude from the Trains Data") %>%
  cols_label(gender = "Gender",
             income = "Income", 
             att_end = "End Attitude") %>% 
  tab_source_note(md("Source: Ryan Enos, *Causal Effect of Intergroup Contact on Exclusionary Attitudes*"))
```

&nbsp;

Now that the table structure looks good, we want to format the numbers themselves. Let's add some dollar signs to the income column using `fmt_currency()`. This function also adds commas (if you want commas without dollar signs use `fmt_number()`). The `c()` within `fmt_currency()` denotes the variable being formatted as a currency:


```{r}
x %>% 
  gt() %>%
    tab_header(title = md("**Enos Data Observations**"), 
               subtitle = "Gender, Income, and End Attitude from the Trains Data")%>%
    cols_label(gender = "Gender",
               income = "Income", 
               att_end = "End Attitude") %>% 
    tab_source_note(md("Source: Ryan Enos, *Causal Effect of Intergroup 
                       Contact on Exclusionary Attitudes*")) %>%
    fmt_currency(columns = c(income), 
                 decimals = 0) 
```

&nbsp;

Note that the line return in the title between "Intergroup" and "Contact" does not effect or break up the title displayed by `md()`.

### Regression tables {-}

We can making a **gt** table with a `stan_glm()` regression object. Key to this is the **gtsummary** package and its `tbl_regression()` function.

```{r, message=FALSE}
library(rstanarm)
library(broom.mixed)
library(gtsummary)

fit2 <- stan_glm(att_end ~ party, data = trains, refresh = 0)

tbl_regression(fit2, 
               intercept = TRUE, 
               estimate_fun = function(x) style_sigfig(x, digits = 2)) %>%
  as_gt() %>%
    tab_header(title = "Regression of Attitudes about Immigration", 
               subtitle = "The Effect of Party on End Attitude") %>%
    tab_source_note(md("Source: Ryan Enos, *Causal Effect of Intergroup 
                        Contact on Exclusionary Attitudes*"))
```

&nbsp;




